server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java

```diff
@@ -345,6 +348,15 @@ public Query prefixQuery(String value, @Nullable MultiTermQuery.RewriteMethod me
         throw new QueryShardException(context, "Can only use prefix queries on keyword and text fields - not on [" + name + "] which is of type [" + typeName() + "]");
     }
 
+    public Query wildcardQuery(String value, QueryShardContext context) {
```

Comment:
I think the same reasoning can apply for `wildcard`, `prefix` and `regex` queries so the default impl should throw a `QueryShardException` ? Only `StringFieldType` fields should be able to build a `wildcard` query.

Reply:
The `keyword` field applies the normalizer on `termQuery`. Depending on the normalizer the wildcard and escaped characters could be removed/replaced so I wonder if we should apply the same logic than `QueryParserBase#analyzeWildcard` for `keyword` fields. This is out of scope for this pr but it made me realize that we might have a bug here.

Reply:
> I didn't go further and disallow wildcard queries for all non-keyword or text fields, as some other field types like _index explicitly support wildcard queries.

I missed this part sorry. I think we should explicitly add the support in the `_index` field type rather than supporting this query on all fields. Currently the support for `prefix` queries is also broken so we don't really use this ability.

Reply:
Okay, I'll make sure only string fields support wildcards by default. Maybe I'll add an upgrade note too in case this breaks some types we don't have test coverage for?

Reply:
> This is out of scope for this pr but it made me realize that we might have a bug here.

Makes sense, I'll make a note to follow-up on this.

Reply:
This change would only break `wildcard` query on these fields, right ? +1 to make them string fields, `prefix` and `regex` query do not work currently because of this so it would be a bug fix. I am also ok to do that in a follow up, the changes in this pr have a different scope.

Reply:
👍